dhall/src/Dhall/DirectoryTree.hs

```diff
@@ -102,14 +197,43 @@ toDirectoryTree path expression = case expression of
         Text.IO.writeFile path text
 
     Some value ->
-        toDirectoryTree path value
+        toDirectoryTree allowSeparators path value
 
-    App (Field (Union _) _) value ->
-        toDirectoryTree path value
+    App (Field (Union _) _) value -> do
+        toDirectoryTree allowSeparators path value
 
     App None _ ->
         return ()
 
+    -- If this pattern matches we assume the user wants to use the fixpoint
+    -- approach, hence we typecheck it and output error messages like we would
+    -- do for every other Dhall program.
+    Lam _ (functionBindingVariable -> r) (Lam _ (functionBindingVariable -> make) body) -> do
+        let body' = Core.renote body
+        let expression' = Core.renote expression
+
+        expected' <- case directoryTreeType of
+            Success x -> return x
+            Failure e -> Exception.throwIO e
+
+        _ <- Core.throws $ TypeCheck.typeOf $ Annot expression' expected'
+
+        let expr = rename r "result" $ rename make "make" body'
+
```

Comment:
Instead of using `result` and `make` as reserved names, I find it's easier to `alphaNormalize` the expression and look for `_@0` and `_@1`.  You can find an example of doing this in the `dhallToJSON` function:

https://github.com/dhall-lang/dhall-haskell/blob/8aa54a54ab628bc5c70e7542c5b9d671289ac1c0/dhall-json/src/Dhall/JSON.hs#L416

Not only is it simpler (you don't need to reimplement renaming/substitution), but it also decreases the likelihood of name collisions or name capture bugs.

Reply:
Oh, nice; I changed that.